src/server.rs

```diff
@@ -1320,6 +1325,65 @@ impl ServerRunner {
         });
         Error::verify_good(&status_code)
     }
+
+    /// Runs the server until it's cancelled.
+    ///
+    /// The server is shut down cleanly upon receiving a cancellation token (could be due `SIGINT` or
+    /// `SIGTERM`) at which point the method returns.
+    ///
+    /// # Errors
+    ///
+    /// This fails when the server cannot be started.
+    #[cfg(feature = "tokio")]
+    pub async fn run_with_cancellation_token(
```

Comment:
Wouldn't a simple `canceled: impl Future<Output = ()>>` argument be sufficient here instead of adding a dependency and feature flag?

Even with the cancellation the code needs to `select!` between both futures. Calling `is_cancelled()` periodically is not sufficient if the other future is pending forever.

Reply:
IMHO, on the App service side, its much easier to pass around `CancellationToken`. They are cheaper to clone. I feel when adding `tokio`, adding `tokio_util` is a trivial increase. In the services that i have seen they all seem to use `TaskTracker/CancellationToken` for better ergonomics.

_DISCLAIMER:_ I come from JVM background and have been using Rust only for the past 1 yr.

Reply:
Sorry, for the confusion. I checked the previous code locally and noticed that the `await` is not called with in the loop as I assumed. The `select!` would only make sense as part of the loop.

I'll check this out and try to investigate a simplified version.

Reply:
Please check if this variant works for you: https://github.com/ndchandar/open62541/pull/1

There is even no need for asynchronous code, because all invocations of the C library are synchronous.

Reply:
@sgoll Please also have a look.

Reply:
You could then merge my PR into your branch and the changes will appear here for a final review.